### PR TITLE
Adding CSD specific slide template for export

### PR DESCRIPTION
**Pull request by BoopBoopBeepBoop**
- generalized adding slides
- slight refactoring to remove dead code

Added a CSD specific slide template in the PPT format files, and modularized the creation of the slide files so that arbitrary elements can be added to a slide. This lets us drop the elements in that we want at runtime, but still associate them back to the CSD slide format

### CODE REVIEW

**Comment by BoopBoopBeepBoop** [comment]
My apologies in advance for size, a lot of it is lines of XML (that you can basically ignore)

File: lib/powerpoint/presentation.rb [hunk 1]
```diff
@@ -23,21 +23,65 @@ def add_intro(title, subtitile = nil)
...
+          Powerpoint::Slide::ImageContent.new(idx: 1, image_path: image_path, centered: true, coords: img_coords)
+      ]})
+
+      @slides << slide
     end
 
     def add_text_picture_slide(title, image_path, content = [])
-      @slides << Powerpoint::Slide::TextPicSplit.new(presentation: self, title: title, image_path: image_path, content: content)
+
+      img_coords = { x: 6172200, y: 1825625, cx: 5181600, cy: 4351338 }
```
Comment (sethbro) [inline lib/powerpoint/presentation.rb:58]: What exactly are these numbers?

File: lib/powerpoint/slide/content.rb [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+        File.open("#{extract_path}/ppt/slides/slide#{index}.xml", 'w') do |f|
+          f << render_slide
+        end
+
+      end
+    end
+
+    # renders an image for a PPT slide, snapping it to coordinates if passed.
+    class ImageContent
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]: Ruby idiom is pretty emphatic on one class per file.

File: lib/powerpoint/slide/content.rb [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    class TableContent
+      include Powerpoint::Util
+
+      attr_reader :content, :id, :idx
+      attr_writer :id, :idx
+
+      def initialize(options={})
+        require_arguments [:content, :idx], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]: Didn't take long for you to embrace the meta. Everyone loves their own reflection ;-)

File: lib/powerpoint/slide/content.rb [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+          output += tmp.render_elem
+          tmp = tmp.next
+        end
+        output + footer
+      end
+
+      def render_elem
+        sz = (options[:size] == nil) ? ' ' : " sz=\"#{options[:size]}\""
+        color = (options[:color] == nil) ? '' : "<a:solidFill>
+                                                  <a:srgbClr val=\"#{options[:color]}\"/>
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]: fyi, there's [another string literal syntax](http://www.zenspider.com/Languages/Ruby/QuickRef.html#strings) I find ideal for situations like this:

``` ruby
str = %(In this string, no need to escape "double quotes" or 'single quotes', but can still interpolate #{variables})
```

File: lib/powerpoint/slide/content.rb [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                                                 </a:solidFill>"
+
+        before = '<a:r><a:rPr dirty="0" lang="en-US"' + sz + ' smtClean="0">' + color + '</a:rPr><a:t>'
+        after = '</a:t></a:r>'
+
+        "#{before}#{values.join('')}#{after}"
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]: `join` without arguments will use an empty string.

File: lib/powerpoint/slide/content.rb [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    class TextContent
+      include Powerpoint::Util
+
+      attr_reader :content, :id, :idx, :bullets
+      attr_writer :id, :idx
+
+      def initialize(options={})
+        require_arguments [:content, :idx], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
+        @bullets = false unless @bullets
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]:
``` ruby
@bullets ||= false
```

would be a little more idiomatic.

File: lib/powerpoint/views/content_type.xml.erb [hunk 7]
```diff
@@ -10,21 +10,40 @@
...
+            ContentType="application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml"/>
+  <Override PartName="/ppt/slideLayouts/slideLayout8.xml"
+            ContentType="application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml"/>
+  <Override PartName="/ppt/slideLayouts/slideLayout9.xml"
+            ContentType="application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml"/>
+  <Override PartName="/ppt/slideLayouts/slideLayout10.xml"
+            ContentType="application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml"/>
+  <Override PartName="/ppt/slideLayouts/slideLayout11.xml"
+            ContentType="application/vnd.openxmlformats-officedocument.presentationml.slideLayout+xml"/>
+  <Override PartName="/ppt/slideLayouts/slideLayout12.xml"
```
Comment (sethbro) [inline lib/powerpoint/views/content_type.xml.erb:43]: Are we limiting to a certain number of slides?

File: lib/powerpoint/views/default_slide_rel.xml.erb [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
+  <Relationship Id="rId1" Target="../slideLayouts/slideLayout<%=slide_layout%>.xml" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/slideLayout"/>
+  <% (images || []).each do |image| %>
+  <Relationship Id="<%= image.embed_rid %>" Target="../media/<%= image.image_name %>" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/image"/>
```
Comment (sethbro) [inline lib/powerpoint/views/default_slide_rel.xml.erb]: Nit. indent.

File: lib/powerpoint/views/picture.xml.erb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    <a:picLocks noChangeAspect="1" noGrp="1"/>
+  </p:cNvPicPr>
+  <% if idx > 0 %>
+  <p:nvPr>
+    <p:ph idx="<%= idx %>"/>
+  </p:nvPr>
+  <% else %>
+  <p:nvPr/>
+  <% end %>
+</p:nvPicPr>
```
Comment (sethbro) [inline lib/powerpoint/views/picture.xml.erb]: Nit. Would love some blank lines for quicker scanning of basic doc structure.

File: lib/powerpoint/views/text.xml.erb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    <a:spLocks noGrp="1"/>
+  </p:cNvSpPr>
+  <p:nvPr>
+    <p:ph idx="<%= idx %>"/>
+  </p:nvPr>
+</p:nvSpPr>
+<p:spPr/>
+<p:txBody>
+  <% unless bullets %>
+  <a:bodyPr>
```
Comment (sethbro) [inline lib/powerpoint/views/text.xml.erb]: More of the same nit. Unless it borks the XML somehow, would prefer to see indentation within conditionals in these views. They're so dense already, proper indentation would make it easier to find & follow what logic is present.

File: spec/powerpoint/slide/content_spec.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    expect {
+      str = Powerpoint::Slide::TextContent.new(text_opts).render
+    }.not_to raise_error
+  end
+
+  it 'default slide has content' do
+
+    # img_opts = {
```
Comment (sethbro) [inline spec/powerpoint/slide/content_spec.rb]: Nit. remove.

File: lib/powerpoint/slide/content.rb [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        output = header
+        while tmp != nil
+          output += tmp.render_elem
+          tmp = tmp.next
+        end
+        output + footer
+      end
+
+      def render_elem
+        sz = (options[:size] == nil) ? ' ' : " sz=\"#{options[:size]}\""
```
Comment (sethbro) [inline lib/powerpoint/slide/content.rb]: Can do `options[:size].nil?`

File: lib/powerpoint/presentation.rb (hunk 1 above)
Comment (BoopBoopBeepBoop) [inline lib/powerpoint/presentation.rb:58]: magic numbers coming from the image coordinates in the template file :/ if the templates were explicit datastructures I'd like to just grab them ... but for right now they're not

File: lib/powerpoint/slide/content.rb (hunk 2 above)
Comment (BoopBoopBeepBoop) [inline lib/powerpoint/slide/content.rb]: :(

File: lib/powerpoint/presentation.rb [hunk 13]
```diff
@@ -23,21 +23,65 @@ def add_intro(title, subtitile = nil)
     end
 
     def add_textual_slide(title, content = [])
-      @slides << Powerpoint::Slide::Textual.new(presentation: self, title: title, content: content)
+      slide = Powerpoint::Slide::DefaultSlide.new({
```
Comment (zbelzer) [inline lib/powerpoint/presentation.rb]: Implicit hash notation por favor

File: lib/powerpoint/presentation.rb [hunk 14]
```diff
@@ -23,21 +23,65 @@ def add_intro(title, subtitile = nil)
...
     def add_pictorial_slide(title, image_path, coords = {})
-      @slides << Powerpoint::Slide::Pictorial.new(presentation: self, title: title, image_path: image_path, coords: coords)
+
+      img_coords = { x: 826325, y: 1825625, cx: 10515600, cy: 4351338 }
+
+      slide = Powerpoint::Slide::DefaultSlide.new({
+        title: title,
+        slide_layout: 2,
+        elements: [
+          Powerpoint::Slide::ImageContent.new(idx: 1, image_path: image_path, centered: true, coords: img_coords)
```
Comment (zbelzer) [inline lib/powerpoint/presentation.rb]: Running off the git review page here, I'd put newlines in

File: lib/powerpoint/presentation.rb [hunk 15]
```diff
@@ -70,7 +114,10 @@ def save(path)
     end
 
     def file_types
-      slides.map {|slide| slide.file_type if slide.respond_to? :file_type }.compact.uniq
+      types = slides.map do |slide|
+        slide.file_types if slide.respond_to? :file_types
```
Comment (zbelzer) [inline lib/powerpoint/presentation.rb]: Use parens unless it's a DSL method (like one on Object or the self object in context)

File: lib/powerpoint/slide/content.rb [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+    end
+
+    # renders an image for a PPT slide, snapping it to coordinates if passed.
+    class ImageContent
+      include Powerpoint::Util
+
+      attr_reader :image_name, :coords, :image_path, :id, :idx, :embed_rid
+      attr_writer :id, :idx, :embed_rid
+
+      def initialize(options={})
```
Comment (zbelzer) [inline lib/powerpoint/slide/content.rb]: spaces between arg names and defaults

File: lib/powerpoint/slide/content.rb [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        require_arguments [:image_path], options
+        options.each {|k, v| instance_variable_set("@#{k}", v)}
+        @idx = -1 unless @idx
+        @centered = false unless @centered
+        if !@coords
```
Comment (zbelzer) [inline lib/powerpoint/slide/content.rb]: Try to phrase things positively... ie if @coords blah...

File: lib/powerpoint/slide/content.rb [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+    class Text
+      include Powerpoint::Util
+
+      attr_reader :values, :next, :options
+      attr_writer :next
+
+      def initialize(values, options={})
+        # require_arguments [], options
+        options.each { |k, v| instance_variable_set("@#{k}", v) }
+        if values.instance_of? String
```
Comment (zbelzer) [inline lib/powerpoint/slide/content.rb]: Array(values) does this trick. so @values = Array(values)

File: lib/powerpoint/slide/content.rb [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+            Text.new(@values + [other], options)
+          end
+        else
+          self.copy(other.copy(nil))
+        end
+      end
+
+      def copy(new_next)
+        next_opt = @options.clone
+        if next_opt[:next] != nil
```
Comment (zbelzer) [inline lib/powerpoint/slide/content.rb]: Ruby's boolean system is always !nil, so `if next_opt[:next]`

File: lib/powerpoint/views/content_type.xml.erb (hunk 7 above)
Comment (BoopBoopBeepBoop) [inline lib/powerpoint/views/content_type.xml.erb:43]: no, those are layout types, which are locked at the moment

**Comment by sethbro** [comment]
LGTM. @zbelzer?

**Pull request merged**